hibernate-core/src/main/java/org/hibernate/metamodel/mapping/EmbeddableDiscriminatorConverter.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html.
+ */
+package org.hibernate.metamodel.mapping;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.hibernate.AssertionFailure;
+import org.hibernate.HibernateException;
+import org.hibernate.internal.util.collections.CollectionHelper;
+import org.hibernate.metamodel.mapping.internal.EmbeddableDiscriminatorValueDetailsImpl;
+import org.hibernate.metamodel.model.domain.NavigableRole;
+import org.hibernate.type.BasicType;
+import org.hibernate.type.descriptor.java.JavaType;
+
+/**
+ * Handles conversion of discriminator values for embeddable subtype classes
+ * to their domain typed form.
+ *
+ * @author Marco Belladelli
+ * @see EmbeddableDiscriminatorMapping
+ */
+public class EmbeddableDiscriminatorConverter<O, R> extends DiscriminatorConverter<O, R> {
+	public static <O, R> EmbeddableDiscriminatorConverter<O, R> fromValueMappings(
+			NavigableRole role,
+			JavaType<O> domainJavaType,
+			BasicType<R> underlyingJdbcMapping,
+			Map<Object, String> valueMappings) {
+		final List<DiscriminatorValueDetails> valueDetailsList = CollectionHelper.arrayList( valueMappings.size() );
+		valueMappings.forEach( (value, embeddableClassName) -> valueDetailsList.add( new EmbeddableDiscriminatorValueDetailsImpl(
+				value,
+				embeddableClassName
+		) ) );
+		return new EmbeddableDiscriminatorConverter<>(
+				role,
+				domainJavaType,
+				underlyingJdbcMapping.getJavaTypeDescriptor(),
+				valueDetailsList
+		);
+	}
+
+	private final Map<Object, DiscriminatorValueDetails> discriminatorValueToDetailsMap;
+	private final Map<String, DiscriminatorValueDetails> embeddableClassNameToDetailsMap;
+
+	public EmbeddableDiscriminatorConverter(
+			NavigableRole discriminatorRole,
+			JavaType<O> domainJavaType,
+			JavaType<R> relationalJavaType,
+			List<DiscriminatorValueDetails> valueMappings) {
+		super( discriminatorRole, domainJavaType, relationalJavaType );
+
+		this.discriminatorValueToDetailsMap = CollectionHelper.concurrentMap( valueMappings.size() );
+		this.embeddableClassNameToDetailsMap = CollectionHelper.concurrentMap( valueMappings.size() );
```

Comment:
Regular maps should be fine

Reply:
I remember trying to use simple `HashMap`s and getting some test failures, but now it looks like they work. Maybe it was just a temporary workaround I needed since I used `Class` as keys early on.